Branch creation gives error in Bitbucket

Thanks for the quick release of the YACC plugin version 1.9 for Bitbucket.

Since this update to Bitbucket 4.0.1. and YACC 1.9 we get errors during creation of a branch. We have no "Branch Matching Regexp". 

Error in Bitbucket: 
An error occurred while processing the request. Check the server logs for more information

Error in the server logs:
.a.e.i.AsynchronousAbleEventDispatcher There was an exception thrown trying to dispatch event [com.atlassian.bitbucket.event.branch.BranchCreationRequestedEvent[source=com.atlassian.stash.internal.repository.DefaultRefService@759c8632]] from the invoker [SingleParameterMethodListenerInvoker{method=public void com.isroot.stash.plugin.YaccBranchCreationListener.onBranchCreation(com.atlassian.bitbucket.event.branch.BranchCreationRequestedEvent), listener=com.isroot.stash.plugin.YaccBranchCreationListener@498a6c2e}]

@sford 
I have never seen this exception, but after some investigation, I found that this error occurs under the following conditions. In addition to, checking of 'Branch Name Regex' is ignored.

* Set 'Branch Name Regex' in Hook configuration of the repository (not in global setting).
* Set 'Write' permission to user.
* The user create a branch from Stash-UI.

Full error messages are as follows.
```
2015-09-29 01:48:31,123 ERROR [http-bio-7990-exec-8] user @1RNQTZ8x108x2685x0 cert57 0:0:0:0:0:0:0:1 "POST /rest/branch-utils/latest/projects/PROJECT_1/repos/rep_1/branches HTTP/1.1" c.a.e.i.AsynchronousAbleEventDispatcher There was an exception thrown trying to dispatch event [com.atlassian.stash.branch.BranchCreationRequestedEvent[source=com.atlassian.stash.internal.branch.DefaultBranchService@409edf42]] from the invoker [SingleParameterMethodListenerInvoker{method=public void com.isroot.stash.plugin.YaccBranchCreationListener.onBranchCreation(com.atlassian.stash.branch.BranchCreationRequestedEvent), listener=com.isroot.stash.plugin.YaccBranchCreationListener@7c71af1c}]
java.lang.RuntimeException: You are not permitted to access this resource
	at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:54) ~[atlassian-event-2.3.5.jar:na]
	at com.atlassian.event.internal.AsynchronousAbleEventDispatcher$1$1.run(AsynchronousAbleEventDispatcher.java:48) ~[atlassian-event-2.3.5.jar:na]
	at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:253) [guava-11.0.2-atlassian-02.jar:na]
	at com.atlassian.event.internal.AsynchronousAbleEventDispatcher.dispatch(AsynchronousAbleEventDispatcher.java:107) [atlassian-event-2.3.5.jar:na]
	at com.atlassian.stash.internal.event.LockFreeEventPublisher$Publisher.dispatch(LockFreeEventPublisher.java:202) [stash-platform-3.7.0.jar:na]
	at com.atlassian.stash.internal.event.LockFreeEventPublisher.publish(LockFreeEventPublisher.java:75) [stash-platform-3.7.0.jar:na]
	at com.atlassian.stash.internal.event.TransactionAwareEventPublisher.publish(TransactionAwareEventPublisher.java:90) [stash-platform-3.7.0.jar:na]
	at com.atlassian.stash.internal.branch.DefaultBranchService.raiseBranchCreationRequestedEvent(DefaultBranchService.scala:120) [stash-branch-utils-3.7.0.jar:na]
	at com.atlassian.stash.internal.branch.DefaultBranchService.createBranch(DefaultBranchService.scala:49) [stash-branch-utils-3.7.0.jar:na]
	at com.atlassian.stash.internal.branch.rest.BranchUtilsResource.createBranch(BranchUtilsResource.java:83) [stash-branch-utils-3.7.0.jar:na]
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:25) [applinks-plugin-4.3.5_1413908446000.jar:na]
	at com.atlassian.labs.httpservice.resource.ResourceFilter.doFilter(ResourceFilter.java:59) [httpservice-bridge-0.6.2.jar:na]
	at com.atlassian.stash.internal.spring.security.StashAuthenticationFilter.doFilter(StashAuthenticationFilter.java:86) [StashAuthenticationFilter.class:na]
	at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doInsideSpringSecurityChain(BeforeLoginPluginAuthenticationFilter.java:111) [BeforeLoginPluginAuthenticationFilter.class:na]
	at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:77) [BeforeLoginPluginAuthenticationFilter.class:na]
	at com.atlassian.security.auth.trustedapps.filter.TrustedApplicationsFilter.doFilter(TrustedApplicationsFilter.java:100) [atlassian-trusted-apps-core-3.0.8.jar:na]
	at com.atlassian.oauth.serviceprovider.internal.servlet.OAuthFilter.doFilter(OAuthFilter.java:79) [atlassian-oauth-service-provider-plugin-1.9.10_1420748386000.jar:na]
	at com.atlassian.fastdev.AutoReloadFilter.doFilter(AutoReloadFilter.java:63) [fastdev-plugin-2.5_1443457991333.jar:na]
	at com.atlassian.analytics.client.filter.DefaultAnalyticsFilter.doFilter(DefaultAnalyticsFilter.java:32) [analytics-client-3.56_1421897236000.jar:na]
	at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:32) [analytics-client-3.56_1421897236000.jar:na]
	at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doBeforeBeforeLoginFilters(BeforeLoginPluginAuthenticationFilter.java:89) [BeforeLoginPluginAuthenticationFilter.class:na]
	at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:75) [BeforeLoginPluginAuthenticationFilter.class:na]
	at com.atlassian.stash.internal.request.DefaultRequestManager.doAsRequest(DefaultRequestManager.java:85) [stash-service-impl-3.7.0.jar:na]
	at com.atlassian.stash.internal.hazelcast.ConfigurableWebFilter.doFilter(ConfigurableWebFilter.java:38) [ConfigurableWebFilter.class:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_51]
	at java.lang.Thread.run(Thread.java:744) [na:1.7.0_51]
	... 226 frames trimmed
Caused by: com.atlassian.stash.exception.AuthorisationException: You are not permitted to access this resource
	at com.atlassian.stash.internal.aop.ExceptionRewriteAdvice.afterThrowing(ExceptionRewriteAdvice.java:36) ~[stash-platform-3.7.0.jar:na]
	at com.isroot.stash.plugin.YaccBranchCreationListener.onBranchCreation(YaccBranchCreationListener.java:59) ~[na:na]
	at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:36) ~[atlassian-event-2.3.5.jar:na]
	... 26 common frames omitted
Caused by: org.springframework.security.access.AccessDeniedException: Access is denied
	at org.springframework.security.access.vote.AffirmativeBased.decide(AffirmativeBased.java:83) ~[spring-security-core-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	... 28 common frames omitted
```

It looks like that AccessDeniedException was thrown in the following method call.
```
settings = repositoryHookService.getSettings(repository, hook.getDetails().getKey());
```

I think this problem appears to be solved by the following code. Would you let me know what you think about it?
```
            settings = securityService.withPermission(Permission.REPO_ADMIN, "Get hook configuration").call(
                    new UncheckedOperation<Settings>() {
                        public Settings perform() {
                            return repositoryHookService.getSettings(repository, hook.getDetails().getKey());
                        }
                    });
```

--- a/src/main/java/com/isroot/stash/plugin/YaccBranchCreationListener.java
+++ b/src/main/java/com/isroot/stash/plugin/YaccBranchCreationListener.java
@@ -42,7 +42,7 @@ public class YaccBranchCreationListener {
     public void onBranchCreation(BranchCreationRequestedEvent event) {
         final Repository repository = event.getRepository();
 
-        RepositoryHook hook = securityService.withPermission(Permission.REPO_ADMIN, "Get plugin configuration").call(
+        final RepositoryHook hook = securityService.withPermission(Permission.REPO_ADMIN, "Get plugin configuration").call(
                 new UncheckedOperation<RepositoryHook>() {
                     public RepositoryHook perform() {
                         return repositoryHookService.getByKey(repository, "com.isroot.stash.plugin.yacc:yaccHook");
@@ -56,7 +56,12 @@ public class YaccBranchCreationListener {
             // Repository hook overrides default pre-receive hook configuration
             log.debug("PreReceiveRepositoryHook configured. Use repository configuration.");
 
-            settings = repositoryHookService.getSettings(repository, hook.getDetails().getKey());
+            settings = securityService.withPermission(Permission.REPO_ADMIN, "Get hook configuration").call(
+                    new UncheckedOperation<Settings>() {
+                        public Settings perform() {
+                            return repositoryHookService.getSettings(repository, hook.getDetails().getKey());
+                        }
+                    });
         } else {
             // Repository hook not configured
             log.debug("PreReceiveRepositoryHook not configured.  Use global configuration.");

--- a/src/test/java/ut/com/isroot/stash/plugin/YaccBranchCreationListenerTest.java
+++ b/src/test/java/ut/com/isroot/stash/plugin/YaccBranchCreationListenerTest.java
@@ -43,7 +43,8 @@ public class YaccBranchCreationListenerTest {
     @Mock private Branch branch;
     @Mock private CancelState cancelState;
     @Mock private SecurityService securityService;
-    @Mock private EscalatedSecurityContext escalatedSecurityContext;
+    @Mock private EscalatedSecurityContext escalatedSecurityContextForRepositoryHook;
+    @Mock private EscalatedSecurityContext escalatedSecurityContextForSettings;
     @Mock private PluginSettingsFactory pluginSettingsFactory;
     @Mock private PluginSettings pluginSettings;
     @Mock private SettingsBuilder settingsBuilder;
@@ -65,8 +66,12 @@ public class YaccBranchCreationListenerTest {
 
         //mock hook retrieval
         when(securityService.withPermission(Permission.REPO_ADMIN, "Get plugin configuration"))
-                .thenReturn(escalatedSecurityContext);
-        when(escalatedSecurityContext.call(any(UncheckedOperation.class))).thenReturn(repositoryHook);
+                .thenReturn(escalatedSecurityContextForRepositoryHook);
+        when(escalatedSecurityContextForRepositoryHook.call(any(UncheckedOperation.class))).thenReturn(repositoryHook);
+
+        when(securityService.withPermission(Permission.REPO_ADMIN, "Get hook configuration"))
+                .thenReturn(escalatedSecurityContextForSettings);
+        when(escalatedSecurityContextForSettings.call(any(UncheckedOperation.class))).thenReturn(settings);
     }
 
     @Test

